src/main/java/net/neoforged/neoforge/client/model/generators/template/ElementBuilder.java

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright (c) NeoForged and contributors
+ * SPDX-License-Identifier: LGPL-2.1-only
+ */
+
+package net.neoforged.neoforge.client.model.generators.template;
+
+import com.google.common.base.Preconditions;
+import java.util.LinkedHashMap;
+import java.util.Map;
+import java.util.Set;
+import java.util.function.BiConsumer;
+import java.util.function.Consumer;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import net.minecraft.client.data.models.model.TextureSlot;
+import net.minecraft.client.renderer.block.model.BlockElement;
+import net.minecraft.client.renderer.block.model.BlockElementFace;
+import net.minecraft.core.Direction;
+import net.neoforged.neoforge.client.model.ExtraFaceData;
+import org.jetbrains.annotations.Nullable;
+import org.joml.Vector3f;
+
+public final class ElementBuilder {
+    private Vector3f from = new Vector3f();
+    private Vector3f to = new Vector3f(16, 16, 16);
+    private final Map<Direction, FaceBuilder> faces = new LinkedHashMap<>();
+    @Nullable
+    private RotationBuilder rotation;
+    private boolean shade = true;
+    private int lightEmission = 0;
+    private int color = 0xFFFFFFFF;
+    private int blockLight = 0;
+    private int skyLight = 0;
+    private boolean hasAmbientOcclusion = true;
+
+    private static void validateCoordinate(float coord, char name) {
+        Preconditions.checkArgument(!(coord < -16.0F) && !(coord > 32.0F), "Position " + name + " out of range, must be within [-16, 32]. Found: %d", coord);
+    }
+
+    private static void validatePosition(Vector3f pos) {
+        validateCoordinate(pos.x(), 'x');
+        validateCoordinate(pos.y(), 'y');
+        validateCoordinate(pos.z(), 'z');
+    }
```

Comment:
It would be nice to move these methods to the bottom of the class as they are just noise to someone reading the class to grasp its API surface.